# this script transforms variance and dispersion data and extracts residuals for QTL mapping
# based on ../../varianceQTL/get_residuals.R but with different input and filters
# 12/23/2020 JR


#### IN PROGRESS - UNFINISHED SCRIPT ###################################


library(preprocessCore)

folder <- "SCAIP1-6"
norm <- "qqnorm_mean-corr"
filters <- c("baseline", "relaxed")
indivBatchFilter <- 8

# load the annotation to add chromosomal coordinates:
anno <- read.table("/wsu/home/groups/piquelab/SCAIP/covariates/gencode.v34lift37.basic.annotation.gff3.gz",header=F,stringsAsFactors=F)
anno$gene_id <- gsub("ID=","",anno$V9)
anno$gene_id <- gsub(";.*","",anno$gene_id)
# remove the _PAR_Y versions:
anno <- anno[-grep("PAR_Y",anno$gene_id),]
## anno$g.id <- gsub("[.].*","",anno$gene_id) # NOT for SCAIP1-6!
library(dplyr)
anno <- filter(anno,V3=="gene")
rownames(anno) <- anno$gene_id
# remove non-autosomal chromosomes:
annoau <- anno[gsub("chr","",anno[,1]) %in% c(1:22),]
# subset to protein-coding only:
annoau <- annoau[grep("gene_type=protein_coding",annoau$V9),]

# for each filter, process the data:
for(filter in filters){

if(filter=="baseline"){
    filterDir <- "tmp7"
}
if(filter=="relaxed"){
    filterDir <- "tmp9"
}

# get data generated by Julong:
# dispersion:
# SCAIP1-6 mean-corrected:
if(norm=="qqnorm_mean-corr"){
load(paste0("/nfs/rprdata/julong/SCAIP/analyses/SCAIP-ALL-2020.03.23/10_RNA.Variance_output/",filterDir,"/1_RNA.PhxNew.RData"))
Phx <- PhxNew2
}

## # change all 0s to NAs:
## Phx[Phx==0] <- NA
 
## # drop genes with >80% NAs:
## Phx <- Phx[rowSums(is.na(Phx))<0.8*ncol(Phx),]
    
## colnames(Phx) <- gsub("_SCAIP.*","",colnames(Phx))

# add gene coordinates:
Phx <- Phx[rownames(Phx) %in% rownames(annoau),]
## Vx <- Vx[rownames(Vx) %in% rownames(annoau),]
annobed <- annoau[,c(1,4,5,7,10)]
colnames(annobed) <- c("Chr","min","max","strand","ID")
annobed$Chr <- gsub("chr","",annobed$Chr)
# make the TSS the start:
annobed[annobed$strand=="+","start"] <- annobed[annobed$strand=="+","min"]
annobed[annobed$strand=="-","start"] <- annobed[annobed$strand=="-","max"]
annobed[annobed$strand=="-","end"] <- annobed[annobed$strand=="-","min"]
annobed[annobed$strand=="+","end"] <- annobed[annobed$strand=="+","max"]
# keep only relevant comumns:
annobed <- annobed[,c("Chr","start","end","ID")]
colnames(annobed)[1] <-"#Chr"
## Vx <- Vx[rownames(Vx) %in% rownames(annoau),]

# load covariate file:
cvall <- read.table("../../covariates/SCAIP1-6_ALOFT_cv.txt", sep="\t", header=T, comment="", quote='"', stringsAsFactors=F)
# important! order rows in cv and colnames in data the same!
rownames(cvall) <- cvall$dbgap.ID

# go through the cell types:
clusters <- unique(gsub("_.*","",colnames(Phx)))
# and treatments:
trts <- c("_CTRL", "_PHA-DEX","_PHA-EtOH","_LPS-DEX","_LPS-EtOH")

#loop through all the trts:
for(t in trts){
        treat <- Phx[,grep(t,colnames(Phx))]
            # remove the "_CTRL":
            colnames(treat) <- gsub(t,"",colnames(treat))
            for(cn in clusters){
                   cell <- treat[,grep(cn,colnames(treat))]
                   #remove cell name from colnames:
                   colnames(cell) <- gsub(paste0(cn,"_"),"",colnames(cell))
                   # save log- and qqnormed data:
                   ## log <- log(cell+0.0001)
                   qnorm <- normalize.quantiles(cell)
                   colnames(qnorm) <- colnames(cell)
                   rownames(qnorm) <- rownames(cell)
                   # extract residuals
                   cv <- cvall[colnames(cell),]
                   # baseline model:
                   design <- model.matrix(~0+ cv$Batch + cv$Sex + cv$cage1 + cv$SCAIP1_6_genPC1 + cv$SCAIP1_6_genPC2 + cv$SCAIP1_6_genPC3)
                   #get residuals:
                   X <- design
                   H <- X %*% solve(t(X) %*% X) %*% t(X)
                   dim(H)
                   He <- (diag(rep(1,ncol(H)))-H)
                   ## Reslog <- as.matrix(log) %*% He
                   Resqnorm <- as.matrix(qnorm) %*% He
                   sum(abs(t(He)-He)) #Should be almost 0
                   #save the residuals:
                   ## Reslog <- data.frame(Reslog)
                   Resqnorm <- data.frame(Resqnorm)
                   ## colnames(Reslog) <- gsub("[.]","-",colnames(log))
                   colnames(Resqnorm) <- gsub("[.]","-",colnames(qnorm))
                   # add coordinates:
                   ## logbed <- cbind(annoau[rownames(Reslog),],Reslog)
                   qnormbed <- cbind(annobed[rownames(Resqnorm),],Resqnorm)
                   # sort in order:
                   ## logbed[,1] <- as.numeric(logbed[,1])
                   ## logbed[,2] <- as.numeric(logbed[,2])
                   qnormbed[,1] <- as.numeric(qnormbed[,1])
                   qnormbed[,2] <- as.numeric(qnormbed[,2])
                   ## logbed <- logbed[order(logbed[,1],logbed[,2]),]
                   qnormbed <- qnormbed[order(qnormbed[,1],qnormbed[,2]),]
                   ## write.table(logbed,paste0("./data/dispersion/log/residuals/",cn, t,"_dispersion.bed"),sep="\t", col.names=T, quote=F, row.names=F)
                   write.table(qnormbed,paste0("./data/",folder,"/dispersion/",filter,"-cell-filter/",norm,"/residuals/",cn, t,"_dispersion.bed"), sep="\t", col.names=T, quote=F,row.names=F)
                   system(paste0("bgzip -f data/",folder,"/dispersion/",filter,"-cell-filter/",norm,"/residuals/",cn, t,"_dispersion.bed"))
                   system(paste0("tabix -p bed data/",folder,"/dispersion/",filter,"-cell-filter/",norm,"/residuals/",cn, t,"_dispersion.bed.gz"))
                        }
        }
}

### END 12/23/2020
